File: leetcode/other/1_two_num.py
from typing import List
import unittest


class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        dirt = {}
        for item, idx in zip(nums, range(len(nums))):
            x = target - item
            if x in dirt.keys():
                return [dirt[x],idx]
            dirt[item] = idx


class TestDict(unittest.TestCase):
    def test_fun(self):
        input = [2, 7, 11, 15]
        target = 9
        output = [0, 1]
        answer = Solution().twoSum(input, target)
        self.assertEqual(output, answer)

    def test_fun_1(self):
        input = [3, 2, 4]
        target = 6
        output = [1, 2]
        answer = Solution().twoSum(input, target)
        self.assertEqual(output, answer)

    def test_fun_2(self):
        input = [3, 3]
        target = 6
        output = [0, 1]
        answer = Solution().twoSum(input, target)
        self.assertEqual(output, answer)


if __name__ == '__main__':
    unittest.main()

File: leetcode/other/1_two_num.cpp
//
// Created by arron on 19-4-26.
//
#include <iostream>
#include <vector>
#include <gtest/gtest.h>
#include <map>

using namespace std;

class Solution {
public:
    vector<int> twoSum(vector<int> &nums, int target) {
        map<int, int> dirt;
        for (int i = 0; i < nums.size(); i++) {
            int x = target - nums[i];
            if (dirt.find(x) == dirt.end()) {
                dirt[nums[i]] = i;
            } else {
                return vector<int>{dirt[x], i};
            }
        }
        return vector<int>();
    }
};

TEST(twoNum, 1) {
    std::cout << std::endl;
    vector<int> input = {2, 7, 11, 15};
    int target = 9;
    vector<int> answer = {0, 1};
    vector<int> output = Solution().twoSum(input, target);
    EXPECT_EQ(answer, output);
}